### PR TITLE
feat: add overview and segment query api

### PR DESCRIPTION
**Pull request by 0xff-dev**
feat: add overview and segment query api.
Fix: https://github.com/bestchains/bc-explorer/issues/32

### CODE REVIEW

File: pkg/viewer/overview.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	End   int64 `json:"end"`
+	Count int64 `json:"count"`
+}
+
+type Overview interface {
+	// Summary returns block height, number of transactions, number of nodes, total number of contracts.
+	Summary() (SummaryResp, error)
+
+	// QueryBySeg query the total number of transactions or blocks for a number of time periods
+	// from, interval,number of time periods
```
Comment (bjwswang) [inline pkg/viewer/overview.go:47]: have you sync with @Jandiasnow  with theses arguments ?

File: pkg/viewer/overview.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+func (o *overview) Summary() (SummaryResp, error) {
+	var resp SummaryResp
+	q := o.db.Model((*models.Transaction)(nil))
+	err := q.ColumnExpr(`count(*) as "txCount"`).Select(&resp.TxCount)
+	if err != nil {
+		return resp, err
+	}
+	err = o.db.Model((*models.Block)(nil)).ColumnExpr(`max("blockNumber") as "blockNumber"`).Select(&resp.BlockNumber)
```
Comment (bjwswang) [inline pkg/viewer/overview.go:66]: keep same as `transaction query`.
use a variable or keep one line just like this

File: pkg/viewer/overview.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		return resp, err
+	}
+	return resp, nil
+}
+
+func (o *overview) QueryBySeg(from, interval, number int64, which string) ([]BySegResp, error) {
+	// If there are more types, cast them into interface implementations
+	f, ok := bySegFuncs[which]
```
Comment (bjwswang) [inline pkg/viewer/overview.go:75]: Good design!

File: pkg/viewer/overview_query_by_seg.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+type BySegFunc func(*pg.DB, int64, int64, int64) ([]BySegResp, error)
+
+var bySegFuncs = map[string]BySegFunc{
+	BlockAggregation:       QueryBlocks,
+	TransactionAggregation: QueryTrnasactions,
+}
+
+// QueryByBlocks
+// from=0,interval=5,number=2
```
Comment (bjwswang) [inline pkg/viewer/overview_query_by_seg.go:32]: interval has unit `Seconds`,right ?

Comment (0xff-dev) [reply pkg/viewer/overview_query_by_seg.go:32]: Yes.

File: pkg/viewer/overview.go (hunk 1 above)
Comment (0xff-dev) [reply pkg/viewer/overview.go:47]: Synced.

**Comment by 0xff-dev** [comment]
@Jandiasnow Do the segmentation query parameters for blocks and transactions need to be adjusted?

File: pkg/viewer/overview_query_by_seg.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	result[index].Count = blockCount
+	return result, nil
+}
+
+func QueryTrnasactions(db *pg.DB, network string, from, interval, number int64) ([]BySegResp, error) {
+	start := from - interval
+	end := from + number*interval
+	result := make([]BySegResp, number+1)
+	transactions := make([]models.Transaction, 0)
+	if err := db.Model(&transactions).Where(`"network"=?`, network).
```
[the remark this answers is not in the document]
Comment (0xff-dev) [reply pkg/viewer/overview_query_by_seg.go:70]: You mean we should get the total number of data in each interval by multiple queries, right?

File: pkg/viewer/overview_query_by_seg.go [hunk 6]
```diff
@@ -33,61 +36,56 @@ var bySegFuncs = map[string]BySegFunc{
...
-		if block.CreatedAt < end {
-			blockCount++
-			continue
-		}
-		result[index].Count = blockCount
-		blockCount = 1
-		end, index = end+interval, index+1
+	wg.Wait()
+	if len(ch) > 0 {
+		return nil, <-ch
```
Comment (bjwswang) [inline pkg/viewer/overview_query_by_seg.go:60]: 666

**Review by bjwswang — APPROVED** [review]
/lgtm

**Pull request merged**